homeassistant/components/modbus/switch.py

```diff
@@ -20,144 +16,91 @@
-            discovery_info[CONF_SWITCHES].extend(config[CONF_REGISTERS])
-        for entry in discovery_info[CONF_SWITCHES]:
-            if CALL_TYPE_COIL in entry:
-                entry[CONF_ADDRESS] = entry[CALL_TYPE_COIL]
-                entry[CONF_INPUT_TYPE] = CALL_TYPE_COIL
-                del entry[CALL_TYPE_COIL]
-            if CONF_REGISTER in entry:
-                entry[CONF_ADDRESS] = entry[CONF_REGISTER]
-                del entry[CONF_REGISTER]
-                if CONF_REGISTER_TYPE in entry:
-                    entry[CONF_INPUT_TYPE] = entry[CONF_REGISTER_TYPE]
-                    del entry[CONF_REGISTER_TYPE]
-            if CONF_SCAN_INTERVAL not in entry:
-                entry[CONF_SCAN_INTERVAL] = DEFAULT_SCAN_INTERVAL
-        config = None
-
     for entry in discovery_info[CONF_SWITCHES]:
-        if CONF_HUB in entry:
-            # from old config!
-            hub: ModbusHub = hass.data[MODBUS_DOMAIN][entry[CONF_HUB]]
-        else:
-            hub: ModbusHub = hass.data[MODBUS_DOMAIN][discovery_info[CONF_NAME]]
-        if entry[CONF_INPUT_TYPE] == CALL_TYPE_COIL:
-            switches.append(ModbusCoilSwitch(hub, entry))
-        else:
-            switches.append(ModbusRegisterSwitch(hub, entry))
+        hub: ModbusHub = hass.data[MODBUS_DOMAIN][discovery_info[CONF_NAME]]
+        switches.append(ModbusSwitch(hub, entry))
     async_add_entities(switches)
 
 
-class ModbusBaseSwitch(SwitchEntity, RestoreEntity, ABC):
+class ModbusSwitch(SwitchEntity, RestoreEntity):
     """Base class representing a Modbus switch."""
 
-    def __init__(self, hub: ModbusHub, config: dict[str, Any]):
+    def __init__(self, hub: ModbusHub, config: dict):
         """Initialize the switch."""
         self._hub: ModbusHub = hub
         self._name = config[CONF_NAME]
         self._slave = config.get(CONF_SLAVE)
         self._is_on = None
         self._available = True
         self._scan_interval = timedelta(seconds=config[CONF_SCAN_INTERVAL])
+        self._address = config[CONF_ADDRESS]
+        if config[CONF_WRITE_TYPE] == CALL_TYPE_COIL:
+            self._write_func = self._hub.write_coil
+            self._command_on = 0x01
+            self._command_off = 0x00
+        else:
+            self._write_func = self._hub.write_register
+            self._command_on = config[CONF_COMMAND_ON]
+            self._command_off = config[CONF_COMMAND_OFF]
+        if CONF_VERIFY in config:
+            self._verify_active = True
+            self._verify_address = config[CONF_VERIFY].get(
+                CONF_ADDRESS, config[CONF_ADDRESS]
+            )
+            self._verify_type = config[CONF_VERIFY].get(
+                CONF_INPUT_TYPE, config[CONF_WRITE_TYPE]
+            )
+            self._state_on = config[CONF_VERIFY].get(CONF_STATE_ON, self._command_on)
+            self._state_off = config[CONF_VERIFY].get(CONF_STATE_OFF, self._command_off)
+
+            if self._verify_type == CALL_TYPE_REGISTER_HOLDING:
```

Comment:
We could create a dict that maps verify type to hub attribute name and use `getattr` to get the correct hub attribute. It's not a super long if/elif/else trail so I think it's optional to change it.

Reply:
Somebody told me that "getattr" is expensive and should be avoided, but I believe it makes the code easier to read, let me change that.